fastapi/dependencies/utils.py

```diff
@@ -243,29 +243,69 @@ def is_scalar_sequence_field(field: ModelField) -> bool:
 
 
 def get_typed_signature(call: Callable[..., Any]) -> inspect.Signature:
+    origin = get_origin(call)
+    provided_type: Dict[str, Any] = {}
+    if origin is not None:
+        provided_type = _get_provided_type_generic(origin, call)
+        call = getattr(origin, "__init__")
     signature = inspect.signature(call)
     globalns = getattr(call, "__globals__", {})
     typed_params = [
         inspect.Parameter(
             name=param.name,
             kind=param.kind,
             default=param.default,
-            annotation=get_typed_annotation(param, globalns),
+            annotation=get_typed_annotation(param, provided_type, globalns),
         )
         for param in signature.parameters.values()
+        if param.name != "self"
     ]
     typed_signature = inspect.Signature(typed_params)
     return typed_signature
 
 
-def get_typed_annotation(param: inspect.Parameter, globalns: Dict[str, Any]) -> Any:
+def _get_provided_type_generic(
+    origin: Callable[..., Any], call: Callable[..., Any]
+) -> Dict[str, Any]:
+    provided_type = {}
+    if hasattr(origin, "__orig_bases__"):
+        base = getattr(origin, '__orig_bases__')[0]
+        if hasattr(base, "__parameters__"):
+            base_parameters = getattr(base, "__parameters__", ())
+            args = get_args(call)
+            provided_type = dict(zip(base_parameters, args))
+    return provided_type
+
+
+def get_typed_annotation(
+    param: inspect.Parameter, provided_type: Dict[str, Any], globalns: Dict[str, Any]
+) -> Any:
     annotation = param.annotation
     if isinstance(annotation, str):
         annotation = ForwardRef(annotation)
         annotation = evaluate_forwardref(annotation, globalns, globalns)
+    annotation_generic = _get_annotation_for_generic(annotation, provided_type)
+    if annotation_generic is not None:
+        annotation = annotation_generic
     return annotation
 
 
+def _get_annotation_for_generic(annotation: Any, provided_type: Dict[str, Any]) -> Type:
+    annotation_args = get_args(annotation)
+    updated_args_annotation = False
+    if annotation in provided_type:
+        annotation = provided_type[annotation]
+    elif annotation_args is not None:
+        args_to_replace = []
+        for arg in annotation_args:
+            if arg in provided_type:
+                arg = provided_type[arg]
+                updated_args_annotation = True
+            args_to_replace.append(arg)
+        if updated_args_annotation:
+            setattr(annotation, "__args__", tuple(args_to_replace))
+    return annotation
+
```

Comment:
Decrease the indentation level if it is possible.

```py
if annotation in provided_type:
    annotation = provided_type[annotation]

annotation_args = get_args(annotation)
if annotation_args is None:
    return annotation

args_to_replace = []
for arg in annotation_args:
    arg = provided_type.get(arg, arg)
    args_to_replace.append(arg)

updated_args_annotation = any(
    arg in provided_type for arg in annotation_args
)
if updated_args_annotation:
    setattr(annotation, "__args__", tuple(args_to_replace))

return annotation
```
In my view this is better.

Reply:
Thanks for your feedback @cikay.

Yes it's clearer but we go through the loop twice. it's seem less efficient to me, what do you think ?

Reply:
Then you can merge the loops. It would still be less indentation level. by the way the built-in function `any` is break the loop for the first `True` value. It is efficient.

Reply:
There is nothing to discuss to apply the early return pattern to decrease the indentation level.